src/Bedrock.Framework.Experimental/Protocols/Kafka/Messaging/KafkaMessageDescriptor.cs

```diff
@@ -0,0 +1,109 @@
+using System.Diagnostics;
+using System.Linq;
+using System;
+using System.Buffers;
+using Bedrock.Framework.Protocols.Kafka.Messaging.SchemaTypes;
+
+namespace Bedrock.Framework.Protocols.Kafka.Messaging
+{
+    /// <summary>
+    /// Defines an API message that can be used as part of the wire protocol between a Kafka client and broker.
+    /// </summary>
+    /// <typeparam name="TRequestSchema">The request schema type.</typeparam>
+    /// <typeparam name="TResponseSchema">The response schema type.</typeparam>
+    /// <remarks>
+    /// A message descriptor brings together a request and response schema type for a version of a Kafka API
+    /// message that is implemented by a Kafka broker. The request and response schema types are inclusive of all
+    /// prior versions of the message up to and including the highest version supported by this protocol implementation.
+    /// The reader and writer implementations provided by a given <see cref="KafkaMessageDescriptor{TRequestSchema, TResponseSchema}"/>
+    /// will ensure forward and backward compatibility by omitting fields or supplying default values for fields
+    /// where approprirate.
+    /// 
+    /// The highest version of a message supported by both Kafka client and broker should be used.
+    /// </remarks>
+    public abstract class KafkaMessageDescriptor<TRequestSchema, TResponseSchema> : IKafkaMessageDescriptor
```

Comment:
Is the idea that you would have a "singleton" collection of these to define message serialisation? I guess you could populate that collection once the broker has told you which versions is supports. Hmmm... 🤔

Reply:
Yes, these would be added to the ServiceCollection as singletons. My idea is that I'll have an `IKafkaMessageDescriptorFactory` that will select the relevant one based on the API metadata returned from the broker. So, given a range of supported versions, the factory will return the highest version that we and the broker support.